internal/fs/fs.go

```diff
@@ -2058,7 +2061,60 @@ func (fs *fileSystem) checkDirNotEmpty(dir inode.BucketOwnedDirInode, name strin
 	return nil
 }
 
-// Rename an old directory to a new directory. If the new directory already
+// Rename an old folder to a new folder in a hierarchical bucket. If the new folder already
+// exists and is non-empty, return ENOTEMPTY. If old folder have open files then return
+// ENOTSUP.
+//
+// LOCKS_EXCLUDED(fs.mu)
+// LOCKS_EXCLUDED(oldParent)
+// LOCKS_EXCLUDED(newParent)
+func (fs *fileSystem) renameFolder(ctx context.Context, oldParent inode.DirInode, oldName string, newParent inode.DirInode, newName string) (err error) {
+	// Set up a function that throws away the lookup count increment from
+	// lookUpOrCreateChildInode (since the pending inodes are not sent back to
+	// the kernel) and unlocks the pending inodes, but only once.
+	var pendingInodes []inode.DirInode
+	defer fs.releaseInodes(&pendingInodes)
+
+	oldDirInode, err := fs.getBucketDirInode(ctx, oldParent, oldName)
+	if err != nil {
+		return err
+	}
+	pendingInodes = append(pendingInodes, oldDirInode)
+
+	if err = fs.ensureNoOpenFilesInDirectory(oldDirInode, oldName); err != nil {
```

Comment:
dont call this method ensureNoOpenFiles. Rename it to ensureNoLocalFilesInDirectory. OpenFiles and localfiles are different. We can have a gcs file which is open as well. you are only checking for local files, so mention that in the method name.

Reply:
Makes sense. Thanks. Done

Reply:
just to confirm here, if we have gcs files open during rename, thats not a problem in rename right? do we have any integration test case for both open files and local files scenarios ? I think having those would be useful.

Reply:
> just to confirm here, if we have gcs files open during rename, thats not a problem in rename right?

Yes. Added unit test. will add e2e test in subsequent PR. 

> do we have any integration test case for both open files and local files scenarios ? I think having those would be useful.

With local files yes. with open GCS file no. will add.